File: CashApp-iOS/CashAppPOS/src/services/PlatformService.ts
/**
 * PlatformService - Service for managing platform-wide settings and configurations
 * Handles communication with the platform settings API while maintaining existing patterns
 */

import SharedDataStore from './SharedDataStore';
import tokenManager from '../utils/tokenManager';

// Base API URL - FIXED: Uses LAN IP for device testing
import API_CONFIG from '../config/api';

const BASE_URL = API_CONFIG.FULL_API_URL;

export interface PlatformSetting {
  key: string;
  value: unknown;
  category: string;
  description: string;
  is_sensitive: boolean;
  updated_at: string | null;
}

export interface PaymentFee {
  percentage: number;
  fixed_fee?: number;
  currency: string;
  high_volume?: {
    threshold: number;
    percentage: number;
    monthly_fee: number;
  };
}

export interface FeatureFlag {
  feature_key: string;
  is_enabled: boolean;
  rollout_percentage: number;
  target_restaurants?: string[];
  description?: string;
}

export interface AuditRecord {
  id: string;
  config_type: string;
  config_key: string;
  entity_id?: string;
  old_value: unknown;
  new_value: unknown;
  change_reason?: string;
  change_source: string;
  changed_by: string;
  changed_at: string;
  ip_address?: string;
  user_agent?: string;
}

export interface FeeCalculation {
  payment_method: string;
  amount: number;
  platform_fee: number;
  restaurant_markup: number;
  effective_fee: number;
  fee_percentage: number;
  currency: string;
}

class PlatformService {
  private static instance: PlatformService;
  private authToken: string | null = null;
  private dataStore: SharedDataStore;

  private constructor() {
    this.loadAuthToken();
    this.dataStore = SharedDataStore.getInstance();
  }

  static getInstance(): PlatformService {
    if (!PlatformService.instance) {
      PlatformService.instance = new PlatformService();
    }
    return PlatformService.instance;
  }

  private async loadAuthToken(): Promise<void> {
    try {
      this.authToken = await tokenManager.getTokenWithRefresh();
    } catch (_error) {
    }
  }

  private async makeRequest(
    endpoint: string,
    method: 'GET' | 'POST' | 'PUT' | 'DELETE' = 'GET',
    data?: unknown,
  ): Promise<unknown> {
    try {
      const url = `${BASE_URL}${endpoint}`;
      const headers: unknown = {
        'Content-Type': 'application/json',
      };

      // Always reload auth token for fresh requests
      await this.loadAuthToken();

      if (this.authToken) {
        headers['Authorization'] = `Bearer ${this.authToken}`;
      }

      const config: RequestInit = {
        method,
        headers,
        timeout: 10000, // 10 second timeout for platform operations
      };

      if (data && (method === 'POST' || method === 'PUT')) {
        config.body = JSON.stringify(_data);
      }

      if (_data) {
      }

      const response = await fetch(_url, config);

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`HTTP ${response.status}: ${response.statusText} - ${errorText}`);
      }

      const result = await response.json();
      return result.data || result;
    } catch (_error) {
      throw error;
    }
  }

  // Platform Settings Management
  async getPlatformSettings(
    category?: string,
    includeSensitive = false,
  ): Promise<PlatformSetting[]> {
    try {
      const params = new URLSearchParams();
      if (_category) {
        params.append('category', category);
      }
      if (_includeSensitive) {
        params.append('include_sensitive', 'true');
      }

      const queryString = params.toString();
      const endpoint = `/platform/settings${queryString ? `?${queryString}` : ''}`;

      const settingsData = await this.makeRequest(_endpoint);

      // Handle different API response formats
      let settingsObject: Record<string, any>;

      if (settingsData && typeof settingsData === 'object') {
        // If it's already an object, use it directly
        settingsObject = settingsData;
      } else {
        // If it's an array or other format, create an empty object
        settingsObject = {};
      }

      // Convert object to array format for easier handling
      return Object.entries(_settingsObject).map(([key, config]: [string, any]) => ({
        key,
        value: config?.value ?? config,
        category: config?.category ?? 'general',
        description: config?.description ?? `Setting for ${key}`,
        is_sensitive: config?.is_sensitive ?? false,
        updated_at: config?.updated_at ?? null,
      }));
    } catch (_error) {
      // Return mock data for demo purposes
      return this.getMockPlatformSettings(_category);
    }
  }

  async getPlatformSetting(configKey: string): Promise<PlatformSetting | null> {
    try {
      const settingData = await this.makeRequest(`/platform/settings/${configKey}`);
      return {
        key: settingData.key,
        value: settingData.value,
        category: settingData.category,
        description: settingData.description,
        is_sensitive: settingData.is_sensitive,
        updated_at: settingData.updated_at,
      };
    } catch (_error) {
      return null;
    }
  }

  async updatePlatformSetting(
    configKey: string,
    configValue: unknown,
    changeReason?: string,
  ): Promise<boolean> {
    try {
      await this.makeRequest(`/platform/settings/${configKey}`, 'PUT', {
        config_value: configValue,
        change_reason: changeReason,
      });
      return true;
    } catch (_error) {
      return false;
    }
  }

  async bulkUpdatePlatformSettings(
    updates: Record<string, any>,
    changeReason?: string,
  ): Promise<{ successful: number; failed: number; errors: Record<string, string> }> {
    try {
      const result = await this.makeRequest('/platform/settings/bulk-update', 'POST', {
        updates,
        change_reason: changeReason,
      });

      return {
        successful: result.successful_updates || 0,
        failed: result.failed_updates || 0,
        errors: result.errors || {},
      };
    } catch (_error) {

      // If the bulk endpoint fails, try individual updates as fallback
      let successful = 0;
      let failed = 0;
      const errors: Record<string, string> = {};

      for (const [key, value] of Object.entries(_updates)) {
        try {
          const success = await this.updatePlatformSetting(_key, value, changeReason);
          if (_success) {
            successful++;
          } else {
            failed++;
            errors[key] = 'Update failed';
          }
        } catch (_error) {
          failed++;
          errors[key] = error.message || 'Unknown error';
        }
      }

      return { successful, failed, errors };
    }
  }

  // Payment Fee Management
  async getPaymentFees(): Promise<Record<string, PaymentFee>> {
    try {
      return await this.makeRequest('/platform/payment-fees');
    } catch (_error) {
      // Return mock data for demo
      return this.getMockPaymentFees();
    }
  }

  async calculatePaymentFee(
    paymentMethod: string,
    amount: number,
    restaurantId?: string,
    monthlyVolume?: number,
  ): Promise<FeeCalculation> {
    try {
      const params = new URLSearchParams({
        amount: amount.toString(),
      });
      if (_restaurantId) {
        params.append('restaurant_id', restaurantId);
      }
      if (_monthlyVolume) {
        params.append('monthly_volume', monthlyVolume.toString());
      }

      return await this.makeRequest(
        `/platform/payment-fees/calculate?${params.toString()}`,
        'POST',
      );
    } catch (_error) {
      // Return mock calculation
      return this.getMockFeeCalculation(_paymentMethod, amount);
    }
  }

  // Feature Flag Management
  async getFeatureFlags(restaurantId?: string): Promise<Record<string, boolean>> {
    try {
      const params = restaurantId ? `?restaurant_id=${restaurantId}` : '';
      return await this.makeRequest(`/platform/feature-flags${params}`);
    } catch (_error) {
      return this.getMockFeatureFlags();
    }
  }

  async updateFeatureFlag(
    featureKey: string,
    isEnabled: boolean,
    rolloutPercentage?: number,
    targetRestaurants?: string[],
  ): Promise<boolean> {
    try {
      await this.makeRequest(`/platform/feature-flags/${featureKey}`, 'PUT', {
        is_enabled: isEnabled,
        rollout_percentage: rolloutPercentage,
        target_restaurants: targetRestaurants,
      });
      return true;
    } catch (_error) {
      return false;
    }
  }

  // Restaurant Settings Management
  async getRestaurantEffectiveSettings(
    restaurantId: string,
    category?: string,
  ): Promise<Record<string, any>> {
    try {
      const params = category ? `?category=${category}` : '';
      return await this.makeRequest(
        `/platform/restaurants/${restaurantId}/effective-settings${params}`,
      );
    } catch (_error) {
      return {};
    }
  }

  async setRestaurantOverride(
    restaurantId: string,
    configKey: string,
    overrideValue: unknown,
    requiresApproval = false,
  ): Promise<boolean> {
    try {
      await this.makeRequest(
        `/platform/restaurants/${restaurantId}/overrides/${configKey}`,
        'PUT',
        {
          override_value: overrideValue,
          requires_approval: requiresApproval,
        },
      );
      return true;
    } catch (_error) {
      return false;
    }
  }

  // Audit Trail
  async getAuditTrail(configKey?: string, entityId?: string, limit = 100): Promise<AuditRecord[]> {
    try {
      const params = new URLSearchParams({ limit: limit.toString() });
      if (_configKey) {
        params.append('config_key', configKey);
      }
      if (_entityId) {
        params.append('entity_id', entityId);
      }

      const result = await this.makeRequest(`/platform/audit-trail?${params.toString()}`);
      return result.audit_records || [];
    } catch (_error) {
      return [];
    }
  }

  // Platform Configuration Sync (for mobile apps)
  async syncPlatformConfig(
    restaurantId?: string,
    categories?: string[],
  ): Promise<{
    platform_settings: Record<string, any>;
    feature_flags: Record<string, boolean>;
    effective_settings: Record<string, any>;
    sync_timestamp: string;
  }> {
    try {
      const params = new URLSearchParams();
      if (_restaurantId) {
        params.append('restaurant_id', restaurantId);
      }
      if (_categories) {
        params.append('categories', categories.join(','));
      }

      return await this.makeRequest(`/platform/sync/platform-config?${params.toString()}`);
    } catch (_error) {
      return {
        platform_settings: {},
        feature_flags: {},
        effective_settings: {},
        sync_timestamp: new Date().toISOString(),
      };
    }
  }

  // Initialization
  async initializeDefaultSettings(): Promise<boolean> {
    try {
      await this.makeRequest('/platform/initialize-defaults', 'POST');
      return true;
    } catch (_error) {
      return false;
    }
  }

  // Mock data methods for demo/fallback purposes
  private getMockPlatformSettings(category?: string): PlatformSetting[] {
    const allSettings: PlatformSetting[] = [
      {
        key: 'payment.fees.qr_code',
        value: { percentage: 1.2, currency: 'GBP' },
        category: 'payment_fees',
        description: 'QR Code payment processing fee',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'payment.fees.stripe',
        value: { percentage: 1.4, fixed_fee: 0.2, currency: 'GBP' },
        category: 'payment_fees',
        description: 'Stripe payment processing fee',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'payment.fees.square',
        value: { percentage: 1.75, currency: 'GBP' },
        category: 'payment_fees',
        description: 'Square payment processing fee',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'payment.fees.sumup',
        value: {
          standard: { percentage: 1.95 },
          high_volume: { threshold: 2714, percentage: 0.95, monthly_fee: 39 },
          currency: 'GBP',
        },
        category: 'payment_fees',
        description: 'SumUp payment processing fee',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'security.max_login_attempts',
        value: 5,
        category: 'security',
        description: 'Maximum login attempts before lockout',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'security.session_timeout',
        value: 3600,
        category: 'security',
        description: 'Session timeout in seconds',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'business.max_discount_percentage',
        value: 50.0,
        category: 'business',
        description: 'Maximum discount percentage allowed',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
      {
        key: 'features.new_payment_ui',
        value: { enabled: false, rollout_percentage: 25.0 },
        category: 'features',
        description: 'New payment interface rollout',
        is_sensitive: false,
        updated_at: '2024-06-22T10:30:00Z',
      },
    ];

    return category ? allSettings.filter(s => s.category === category) : allSettings;
  }

  private getMockPaymentFees(): Record<string, PaymentFee> {
    return {
      qr_code: { percentage: 1.2, currency: 'GBP' },
      stripe: { percentage: 1.4, fixed_fee: 0.2, currency: 'GBP' },
      square: {
        percentage: 1.75,
        currency: 'GBP',
        // Additional Square fee structures
        high_volume: { threshold: 0, percentage: 1.75, monthly_fee: 0 }, // No monthly fee
      },
      sumup: {
        percentage: 1.95,
        currency: 'GBP',
        high_volume: { threshold: 2714, percentage: 0.95, monthly_fee: 39 },
      },
    };
  }

  private getMockFeeCalculation(paymentMethod: string, amount: number): FeeCalculation {
    const fees = this.getMockPaymentFees();
    const feeConfig = fees[paymentMethod];

    if (!feeConfig) {
      throw new Error(`Unknown payment method: ${paymentMethod}`);
    }

    const platformFee = (amount * feeConfig.percentage) / 100 + (feeConfig.fixed_fee || 0);

    return {
      payment_method: paymentMethod,
      amount,
      platform_fee: platformFee,
      restaurant_markup: 0,
      effective_fee: platformFee,
      fee_percentage: (platformFee / amount) * 100,
      currency: feeConfig.currency,
    };
  }

  // Service Charge Configuration
  async getServiceChargeConfig(): Promise<{
    enabled: boolean;
    rate: number;
    description: string;
  }> {
    try {
      return await this.dataStore.getServiceChargeConfig();
    } catch (_error) {
      throw error;
    }
  }

  async updateServiceChargeConfig(
    enabled: boolean,
    rate: number,
    description?: string,
  ): Promise<boolean> {
    try {
        enabled,
        rate,
        description,
      });

      const config = {
        enabled,
        rate,
        description: description || `Platform service charge of ${rate}%`,
        lastUpdated: new Date().toISOString(),
      };

      await this.dataStore.setServiceChargeConfig(_config);
      return true;
    } catch (_error) {
      return false;
    }
  }

  private getMockFeatureFlags(): Record<string, boolean> {
    return {
      new_payment_ui: false,
      enhanced_analytics: true,
      mobile_app_v2: false,
      advanced_reporting: true,
      restaurant_chat: false,
    };
  }
}

export default PlatformService;

File: backend/app/services/platform_service.py
"""
Platform Settings Service
Manages centralized platform configurations and restaurant overrides
"""

import logging
from datetime import datetime
from typing import Dict, Any, Optional, List
from sqlalchemy.orm import Session
from sqlalchemy import and_
from jsonschema import validate, ValidationError

from app.models.platform_config import (
    PlatformConfiguration,
    RestaurantOverride,
    ConfigurationAudit,
    PlatformFeatureFlag,
    DEFAULT_PLATFORM_CONFIGS,
    DEFAULT_FEATURE_FLAGS,
)

logger = logging.getLogger(__name__)


class PlatformSettingsService:
    """Service for managing platform-controlled settings"""

    def __init__(self, db: Session):
        self.db = db

    async def get_platform_setting(self, config_key: str) -> Optional[Dict[str, Any]]:
        """Get a specific platform setting by key"""
        setting = (
            self.db.query(PlatformConfiguration)
            .filter(
                and_(
                    PlatformConfiguration.config_key == config_key,
                    PlatformConfiguration.is_active == True,
                )
            )
            .first()
        )

        if not setting:
            return None

        return {
            "key": setting.config_key,
            "value": setting.config_value,
            "category": setting.category,
            "description": setting.description,
            "is_sensitive": setting.is_sensitive,
            "updated_at": (
                setting.updated_at.isoformat() if setting.updated_at else None
            ),
        }

    async def get_platform_settings(
        self, category: Optional[str] = None, include_sensitive: bool = False
    ) -> Dict[str, Any]:
        """Get all platform settings, optionally filtered by category"""

        query = self.db.query(PlatformConfiguration).filter(
            PlatformConfiguration.is_active == True
        )

        if category:
            query = query.filter(PlatformConfiguration.category == category)

        if not include_sensitive:
            query = query.filter(PlatformConfiguration.is_sensitive == False)

        settings = query.all()

        result = {}
        for setting in settings:
            result[setting.config_key] = {
                "value": setting.config_value,
                "category": setting.category,
                "description": setting.description,
                "is_sensitive": setting.is_sensitive,
                "updated_at": (
                    setting.updated_at.isoformat() if setting.updated_at else None
                ),
            }

        return result

    async def update_platform_setting(
        self,
        config_key: str,
        config_value: Any,
        updated_by: str,
        change_reason: Optional[str] = None,
        change_source: str = "api",
    ) -> bool:
        """Update a platform setting with audit trail"""

        # Get existing setting
        setting = (
            self.db.query(PlatformConfiguration)
            .filter(PlatformConfiguration.config_key == config_key)
            .first()
        )

        if not setting:
            logger.error(f"Platform setting '{config_key}' not found")
            return False

        # Validate against schema if defined
        if setting.validation_schema:
            try:
                validate(instance=config_value, schema=setting.validation_schema)
            except ValidationError as e:
                logger.error(f"Validation failed for {config_key}: {e}")
                raise ValueError(f"Invalid configuration value: {e.message}")

        # Store old value for audit
        old_value = setting.config_value

        # Update setting
        setting.config_value = config_value
        setting.updated_at = datetime.utcnow()
        setting.updated_by = updated_by

        # Create audit record
        audit = ConfigurationAudit(
            config_type="platform",
            config_key=config_key,
            entity_id=setting.id,
            old_value=old_value,
            new_value=config_value,
            change_reason=change_reason,
            change_source=change_source,
            changed_by=updated_by,
        )

        self.db.add(audit)
        self.db.commit()

        logger.info(f"Updated platform setting '{config_key}' by user {updated_by}")
        return True

    async def get_restaurant_effective_settings(
        self, restaurant_id: str, category: Optional[str] = None
    ) -> Dict[str, Any]:
        """Get effective settings for a restaurant (platform + overrides)"""

        # Get platform settings
        platform_settings = await self.get_platform_settings(category=category)

        # Get restaurant overrides
        query = self.db.query(RestaurantOverride).filter(
            and_(
                RestaurantOverride.restaurant_id == restaurant_id,
                RestaurantOverride.is_approved == True,
            )
        )

        overrides = query.all()

        # Merge settings
        effective_settings = {}

        # Start with platform settings
        for key, platform_config in platform_settings.items():
            effective_settings[key] = {
                "value": platform_config["value"],
                "source": "platform",
                "category": platform_config["category"],
                "description": platform_config["description"],
                "can_override": self._can_restaurant_override(key),
            }

        # Apply restaurant overrides
        for override in overrides:
            if override.config_key in effective_settings:
                effective_settings[override.config_key][
                    "value"
                ] = override.override_value
                effective_settings[override.config_key]["source"] = "restaurant"
                effective_settings[override.config_key]["override_id"] = str(
                    override.id
                )

        return effective_settings

    async def set_restaurant_override(
        self,
        restaurant_id: str,
        config_key: str,
        override_value: Any,
        created_by: str,
        requires_approval: bool = False,
    ) -> bool:
        """Set a restaurant override for a platform setting"""

        # Check if override is allowed
        if not self._can_restaurant_override(config_key):
            raise ValueError(f"Restaurant overrides not allowed for '{config_key}'")

        # Get platform setting to validate against limits
        platform_setting = await self.get_platform_setting(config_key)
        if not platform_setting:
            raise ValueError(f"Platform setting '{config_key}' not found")

        # Validate override against platform limits
        platform_limit = self._get_platform_limit(config_key)
        if not self._validate_override(config_key, override_value, platform_limit):
            raise ValueError(f"Override value exceeds platform limits")

        # Check for existing override
        existing = (
            self.db.query(RestaurantOverride)
            .filter(
                and_(
                    RestaurantOverride.restaurant_id == restaurant_id,
                    RestaurantOverride.config_key == config_key,
                )
            )
            .first()
        )

        if existing:
            # Update existing override
            old_value = existing.override_value
            existing.override_value = override_value
            existing.updated_at = datetime.utcnow()
            existing.is_approved = not requires_approval
            override_record = existing
        else:
            # Create new override
            override_record = RestaurantOverride(
                restaurant_id=restaurant_id,
                config_key=config_key,
                override_value=override_value,
                platform_limit=platform_limit,
                is_approved=not requires_approval,
                created_by=created_by,
            )
            self.db.add(override_record)
            old_value = None

        # Create audit record
        audit = ConfigurationAudit(
            config_type="restaurant",
            config_key=config_key,
            entity_id=restaurant_id,
            old_value=old_value,
            new_value=override_value,
            change_reason=f"Restaurant override for {config_key}",
            change_source="restaurant_api",
            changed_by=created_by,
        )

        self.db.add(audit)
        self.db.commit()

        logger.info(f"Set restaurant override for '{config_key}' by user {created_by}")
        return True

    async def get_feature_flags(
        self, restaurant_id: Optional[str] = None
    ) -> Dict[str, bool]:
        """Get feature flags, optionally for a specific restaurant"""

        flags = self.db.query(PlatformFeatureFlag).all()

        result = {}
        for flag in flags:
            is_enabled = flag.is_enabled

            # Check if restaurant is in targeted rollout
            if restaurant_id and flag.target_restaurants:
                if restaurant_id in flag.target_restaurants:
                    is_enabled = True
                elif flag.rollout_percentage < 100.0:
                    # Use restaurant ID for consistent rollout
                    import hashlib

                    hash_val = int(
                        hashlib.md5(
                            f"{flag.feature_key}:{restaurant_id}".encode()
                        ).hexdigest(),
                        16,
                    )
                    percentage = (hash_val % 100) + 1
                    is_enabled = percentage <= flag.rollout_percentage

            result[flag.feature_key] = is_enabled

        return result

    async def get_service_charge_config(self) -> Dict[str, Any]:
        """Get the consolidated service charge configuration."""
        config_keys = [
            "platform.service_charge.enabled",
            "platform.service_charge.rate",
            "platform.service_charge.description",
            "platform.service_charge.currency",
        ]

        settings = await self.get_platform_settings(category="service_charge")

        # Fallback to default values if a key is missing, to prevent TypeErrors
        # The default values are defined in DEFAULT_PLATFORM_CONFIGS in platform_config.py
        # This ensures that even if DB initialization failed or a key was somehow deleted,
        # we return a well-structured response.

        default_configs_map = {
            item["config_key"]: item["config_value"]["value"]
            for item in DEFAULT_PLATFORM_CONFIGS
            if item["config_key"] in config_keys
        }

        enabled = (
            settings.get("platform.service_charge.enabled", {})
            .get("value", {})
            .get("value", default_configs_map.get("platform.service_charge.enabled"))
        )
        rate = (
            settings.get("platform.service_charge.rate", {})
            .get("value", {})
            .get("value", default_configs_map.get("platform.service_charge.rate"))
        )
        description = (
            settings.get("platform.service_charge.description", {})
            .get("value", {})
            .get(
                "value", default_configs_map.get("platform.service_charge.description")
            )
        )
        currency = (
            settings.get("platform.service_charge.currency", {})
            .get("value", {})
            .get("value", default_configs_map.get("platform.service_charge.currency"))
        )

        # Ensure correct types, especially for boolean, as get() can return None
        if not isinstance(enabled, bool):
            enabled = default_configs_map.get(
                "platform.service_charge.enabled", False
            )  # Default to False if type is wrong

        return {
            "service_charge": {
                "enabled": enabled,
                "rate": rate,
                "description": description,
                "currency": currency,
            }
        }

    async def update_service_charge_config(
        self,
        config_data: Dict[str, Any],
        updated_by: str,
        change_reason: Optional[str] = None,
    ) -> bool:
        """Update service charge configuration settings."""
        # config_data is expected to be like:
        # {
        #   "enabled": true,
        #   "rate": 12.5,
        #   "description": "Platform service charge",
        #   "currency": "GBP"
        # }

        all_success = True
        for key, value in config_data.items():
            config_key = f"platform.service_charge.{key}"
            try:
                # Wrap the scalar value in a dictionary as per our storage convention
                success = await self.update_platform_setting(
                    config_key=config_key,
                    config_value={
                        "value": value
                    },  # Ensure value is stored in the {'value': ...} structure
                    updated_by=updated_by,
                    change_reason=change_reason or f"Update to service charge {key}",
                    change_source="service_charge_api",
                )
                if not success:
                    all_success = False
                    logger.warning(
                        f"Failed to update service charge setting: {config_key}"
                    )
            except (
                ValueError
            ) as e:  # Catch validation errors from update_platform_setting
                logger.error(f"Validation error updating {config_key}: {e}")
                all_success = False  # Consider it a failure for this key
                raise  # Re-raise to inform the caller (API endpoint) of the bad request
            except Exception as e:
                logger.error(f"Error updating service charge setting {config_key}: {e}")
                all_success = False

        return all_success

    async def update_feature_flag(
        self,
        feature_key: str,
        is_enabled: bool,
        rollout_percentage: Optional[float] = None,
        target_restaurants: Optional[List[str]] = None,
        updated_by: str = None,
    ) -> bool:
        """Update a feature flag"""

        flag = (
            self.db.query(PlatformFeatureFlag)
            .filter(PlatformFeatureFlag.feature_key == feature_key)
            .first()
        )

        if not flag:
            logger.error(f"Feature flag '{feature_key}' not found")
            return False

        flag.is_enabled = is_enabled
        if rollout_percentage is not None:
            flag.rollout_percentage = rollout_percentage
        if target_restaurants is not None:
            flag.target_restaurants = target_restaurants
        flag.updated_at = datetime.utcnow()

        self.db.commit()

        logger.info(f"Updated feature flag '{feature_key}' to {is_enabled}")
        return True

    async def get_payment_fees(self) -> Dict[str, Dict[str, Any]]:
        """Get all payment processing fees"""

        fee_settings = await self.get_platform_settings(category="payment_fees")

        fees = {}
        for key, config in fee_settings.items():
            provider = key.replace("payment.fees.", "")
            fees[provider] = config["value"]

        return fees

    async def calculate_effective_fee(
        self,
        payment_method: str,
        amount: float,
        restaurant_id: Optional[str] = None,
        monthly_volume: Optional[float] = None,
    ) -> Dict[str, Any]:
        """Calculate effective fee for a payment method"""

        # Get platform fee configuration
        fee_config = await self.get_platform_setting(f"payment.fees.{payment_method}")
        if not fee_config:
            raise ValueError(
                f"No fee configuration for payment method '{payment_method}'"
            )

        fee_data = fee_config["value"]

        # Calculate base platform fee
        if payment_method == "sumup" and monthly_volume:
            # SumUp has volume-based pricing
            if monthly_volume >= fee_data.get("high_volume", {}).get("threshold", 2714):
                percentage = fee_data["high_volume"]["percentage"]
                monthly_fee = fee_data["high_volume"]["monthly_fee"]
                # Estimate monthly fee per transaction
                estimated_transactions = monthly_volume / (
                    amount or 50
                )  # Assume avg £50 transaction
                fee_per_transaction = monthly_fee / max(estimated_transactions, 1)
                total_fee = (amount * percentage / 100) + fee_per_transaction
            else:
                percentage = fee_data["standard"]["percentage"]
                total_fee = amount * percentage / 100
        else:
            # Standard percentage + fixed fee
            percentage = fee_data.get("percentage", 0)
            fixed_fee = fee_data.get("fixed_fee", 0)
            total_fee = (amount * percentage / 100) + fixed_fee

        # Check for restaurant markup (if allowed)
        restaurant_markup = 0.0
        if restaurant_id:
            effective_settings = await self.get_restaurant_effective_settings(
                restaurant_id, category="payment_fees"
            )
            markup_key = f"payment.markup.{payment_method}"
            if markup_key in effective_settings:
                restaurant_markup = effective_settings[markup_key]["value"].get(
                    "percentage", 0
                )

        effective_fee = total_fee + (amount * restaurant_markup / 100)

        return {
            "payment_method": payment_method,
            "amount": amount,
            "platform_fee": total_fee,
            "restaurant_markup": restaurant_markup,
            "effective_fee": effective_fee,
            "fee_percentage": (effective_fee / amount * 100) if amount > 0 else 0,
            "currency": fee_data.get("currency", "GBP"),
        }

    def _can_restaurant_override(self, config_key: str) -> bool:
        """Check if restaurants can override a specific platform setting"""

        # Settings that restaurants can override (with limits)
        allowed_overrides = [
            "payment.markup.*",  # Small markup on payment fees
            "business.discount.maximum",  # Within platform limits
            "ui.theme.*",  # UI customization
            "receipt.customization.*",  # Receipt templates
        ]

        # Payment fees and security settings cannot be overridden
        forbidden_overrides = [
            "payment.fees.*",
            "security.*",
            "compliance.*",
            "features.*",
        ]

        # Check forbidden first
        for pattern in forbidden_overrides:
            if self._matches_pattern(config_key, pattern):
                return False

        # Check allowed
        for pattern in allowed_overrides:
            if self._matches_pattern(config_key, pattern):
                return True

        return False

    def _matches_pattern(self, config_key: str, pattern: str) -> bool:
        """Check if config key matches a wildcard pattern"""
        if pattern.endswith("*"):
            prefix = pattern[:-1]
            return config_key.startswith(prefix)
        return config_key == pattern

    def _get_platform_limit(self, config_key: str) -> Optional[Dict[str, Any]]:
        """Get platform-defined limits for restaurant overrides"""

        limits = {
            "payment.markup.qr_code": {"max_percentage": 0.5},
            "payment.markup.stripe": {"max_percentage": 0.3},
            "payment.markup.square": {"max_percentage": 0.3},
            "business.discount.maximum": {"max_percentage": 50.0},
        }

        return limits.get(config_key)

    def _validate_override(
        self,
        config_key: str,
        override_value: Any,
        platform_limit: Optional[Dict[str, Any]],
    ) -> bool:
        """Validate restaurant override against platform limits"""

        if not platform_limit:
            return True

        if config_key.startswith("payment.markup."):
            max_percentage = platform_limit.get("max_percentage", 0)
            value_percentage = override_value.get("percentage", 0)
            return value_percentage <= max_percentage

        if config_key == "business.discount.maximum":
            max_percentage = platform_limit.get("max_percentage", 100)
            return override_value.get("percentage", 0) <= max_percentage

        return True

    async def initialize_default_settings(self) -> bool:
        """Initialize platform with default configurations"""

        try:
            # Add default platform configurations
            for config_data in DEFAULT_PLATFORM_CONFIGS:
                existing = (
                    self.db.query(PlatformConfiguration)
                    .filter(
                        PlatformConfiguration.config_key == config_data["config_key"]
                    )
                    .first()
                )

                if not existing:
                    config = PlatformConfiguration(**config_data)
                    self.db.add(config)

            # Add default feature flags
            for flag_data in DEFAULT_FEATURE_FLAGS:
                existing = (
                    self.db.query(PlatformFeatureFlag)
                    .filter(PlatformFeatureFlag.feature_key == flag_data["feature_key"])
                    .first()
                )

                if not existing:
                    flag = PlatformFeatureFlag(**flag_data)
                    self.db.add(flag)

            self.db.commit()
            logger.info("Default platform settings initialized successfully")
            return True

        except Exception as e:
            logger.error(f"Failed to initialize default settings: {e}")
            self.db.rollback()
            return False

    async def get_audit_trail(
        self,
        config_key: Optional[str] = None,
        entity_id: Optional[str] = None,
        limit: int = 100,
    ) -> List[Dict[str, Any]]:
        """Get configuration audit trail"""

        query = self.db.query(ConfigurationAudit).order_by(
            ConfigurationAudit.changed_at.desc()
        )

        if config_key:
            query = query.filter(ConfigurationAudit.config_key == config_key)

        if entity_id:
            query = query.filter(ConfigurationAudit.entity_id == entity_id)

        audit_records = query.limit(limit).all()

        result = []
        for record in audit_records:
            result.append(
                {
                    "id": str(record.id),
                    "config_type": record.config_type,
                    "config_key": record.config_key,
                    "entity_id": str(record.entity_id) if record.entity_id else None,
                    "old_value": record.old_value,
                    "new_value": record.new_value,
                    "change_reason": record.change_reason,
                    "change_source": record.change_source,
                    "changed_by": str(record.changed_by),
                    "changed_at": record.changed_at.isoformat(),
                    "ip_address": record.ip_address,
                    "user_agent": record.user_agent,
                }
            )

        return result
